lib/utils/qemu.ts

```diff
@@ -84,47 +86,50 @@ export const getQemuPath = function (arch: string) {
 	);
 };
 
-export function installQemu(arch: string) {
-	const request = require('request') as typeof import('request');
-	const fs = require('fs') as typeof import('fs');
-	const zlib = require('zlib') as typeof import('zlib');
-	const tar = require('tar-stream') as typeof import('tar-stream');
-
-	return getQemuPath(arch).then(
-		(qemuPath) =>
-			new Promise(function (resolve, reject) {
-				const installStream = fs.createWriteStream(qemuPath);
-
-				const qemuArch = balenaArchToQemuArch(arch);
-				const fileVersion = QEMU_VERSION.replace(/^v/, '').replace('+', '.');
-				const urlFile = encodeURIComponent(
-					`qemu-${fileVersion}-${qemuArch}.tar.gz`,
-				);
-				const urlVersion = encodeURIComponent(QEMU_VERSION);
-				const qemuUrl = `https://github.com/balena-io/qemu/releases/download/${urlVersion}/${urlFile}`;
-
-				const extract = tar.extract();
-				extract.on('entry', function (header, stream, next) {
-					stream.on('end', next);
-					if (header.name.includes(`qemu-${qemuArch}-static`)) {
-						stream.pipe(installStream);
-					} else {
-						stream.resume();
-					}
-				});
+async function installQemu(arch: string, qemuPath: string) {
+	const qemuArch = balenaArchToQemuArch(arch);
+	const fileVersion = QEMU_VERSION.replace(/^v/, '').replace('+', '.');
+	const urlFile = encodeURIComponent(`qemu-${fileVersion}-${qemuArch}.tar.gz`);
+	const urlVersion = encodeURIComponent(QEMU_VERSION);
+	const qemuUrl = `https://github.com/balena-io/qemu/releases/download/${urlVersion}/${urlFile}`;
 
-				return request(qemuUrl)
-					.on('error', reject)
-					.pipe(zlib.createGunzip())
-					.on('error', reject)
-					.pipe(extract)
-					.on('error', reject)
-					.on('finish', function () {
-						fs.chmodSync(qemuPath, '755');
-						resolve();
-					});
-			}),
-	);
+	const request = await import('request');
+	const fs = await import('fs');
+	const zlib = await import('zlib');
+	const tar = await import('tar-stream');
+
+	const extract = tar.extract();
+	const installStream = fs.createWriteStream(qemuPath);
+	try {
+		extract.on('entry', function (header, stream, next) {
+			stream.on('end', next);
+			if (header.name.includes(`qemu-${qemuArch}-static`)) {
+				stream.pipe(installStream);
+			} else {
+				stream.resume();
+			}
+		});
+
+		await new Promise((resolve, reject) => {
+			request(qemuUrl)
+				.on('error', reject)
+				.pipe(zlib.createGunzip())
+				.on('error', reject)
+				.pipe(extract)
+				.on('error', reject)
+				.on('finish', function () {
+					fs.chmodSync(qemuPath, '755');
+					resolve();
+				});
+		});
+	} catch (err) {
+		try {
+			await fs.promises.unlink(qemuPath);
```

Comment:
The `fs.createWriteStream(qemuPath)` statement towards the beginning of the function creates a zero-length file on disk. On successful download, the file is expanded to the size of the download. On errors, before this PR the file was left behind and then the CLI would assume the file no longer needed to be downloaded (a bug). I've added this `unlink` call to delete "corrupt" files on errors.

Reply:
The commit that this comment referred to was moved to PR #2207.